# Exercise 4.5 Data Consistency Checks

# 1. Importing libraries

In [561]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Importing Data

In [563]:
path = r'C:\Users\Barb\07-2024 Instacart Basket Analysis'

In [564]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original', 'products.csv'), index_col = False)

In [565]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared', 'orders_wrangled.csv'), index_col = False)

# 3. Data Consistency checks

## 3.1 Mixed-Type Data

In [568]:
# Check for mixed-type data in df_prods

In [569]:
for col in df_prods.columns.tolist():
    weird = (df_prods[[col]].map(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_prods[weird]) > 0:
        print(col)       

product_name


In [570]:
df_prods.info() #check data type for all columns in df 
#Mixed-type data occurs when multiple data types exist in the same column.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [571]:
#tried to change data type of product_name to int64, and I got an error
#didn't change any data type

# 4. Missing Values

In [573]:
#check df_prods for missing values

In [574]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [575]:
#view these 16 missing values

In [576]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [577]:
df_nan.head()

product_id product_name  aisle_id  department_id  prices
33           34          NaN       121             14    12.2
68           69          NaN        26              7    11.8
115         116          NaN        93              3    10.8
261         262          NaN       110             13    12.1
525         525          NaN       109             11     1.2

In [578]:
#find out # of rows and columns with missing values
#1st, find out rows of entire df
df_prods.shape

(49693, 5)

In [579]:
#create a new dataframe w/o the missing values

In [580]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [581]:
df_prods_clean.shape

(49677, 5)

# 5. Duplicates

## 5.1 Finding Duplicates

In [584]:
#ID all full duplicates within your dataframe

In [585]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [586]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [587]:
#compare # of rows in df_prods_clean df before and after removal of duplicates

In [588]:
df_prods_clean.shape

(49677, 5)

## 5.2 Deleting Duplicates

In [590]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()      #drop duplicate rows

In [591]:
df_prods_clean_no_dups.shape

(49672, 5)

In [592]:
df_prods_clean.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

# 6. Task 4.5

#### Task #1. See above
#### Task #2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, 
#### share in a markdown cell whether anything about the data looks off or should be investigated further.  Keep eye on min and max values.

In [595]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  hour_time_of_order  days_since_prior_order  
count        3.421083e+06        3.421083e+06            3.214874e+06  
mean         2.776219e+00        1.345202e+01            1.111484e+01  
std          2.046829e+00        4.226088e+00            9.206737e+00  
min          0.000000e+00        0.000000e+00            0.000000e+00  
25%          1.000000e+00        1.000000e+01            4.000000e+00  
50%          3.000000e+00        1.300000e+01            7.000000e+00  
75%          5.000000e+00        1.600000e+01            1.500000e+01  
max          6.000000e+00        2.300000e+01            3.000000e+01

### #2 Answer: Column labeled "Unnamed: 0" is unusual for df_ords dataframe. I need to drop this column

In [597]:
# Drop unnamed columns in df_ords

In [598]:
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [599]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  hour_time_of_order  \
0   2539329        1             1                   2                   8   
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### Task #3. Check for mixed-type data in your df_ords dataframe.

In [601]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)   

### #3 Answer: No mixed type data in df_ords

#### Task #4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.
####    Mixed-type data occurs when multiple data types exist in the same column.

### #4 Answer: There is no mixed-type data in the df_ords

#### Task #5a. Run a check for missing values in your df_ords dataframe.

In [606]:
df_ords.isnull().sum()   #ID'd all of the missing values

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
hour_time_of_order             0
days_since_prior_order    206209
dtype: int64

#### Task #5b. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [608]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]      #View all the missing values by creating a subset of df

In [609]:
df_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         hour_time_of_order  days_since_prior_order  
0                         8                     NaN  
11                       11                     NaN  
26                       14                     NaN  
39                       11                     NaN  
45                       12                     NaN  
...                     ...                     ...  
3420930                  12                     NaN  
3420934                  18                     NaN  
3421002                  19                     NaN  
3421019                  15                     NaN  
3421069                  11                     NaN  

[206209 rows x 6 columns]

### #5b Answer: Out of 3,431,083 total entries, there were 206,209 missing entries of days_since_prior_order.
### I noticed that all of these missing values were associated with the customer's 1st order.

#### Task #6. Address missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

### #6 Answer: I am going to filter out the missing values by making a new subset df: df_ords_clean
### These missing values of days_since_prior_order are not needed, so I removed them.

In [613]:
df_ords.shape  #need to know # of rows in current df to compare it to # of rows in new df and check if missing values were removed

(3421083, 6)

In [614]:
#create new df by filtering out missing values

In [615]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [616]:
df_ords_clean.head()

order_id  user_id  order_number  orders_day_of_week  hour_time_of_order  \
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   
5   3367565        1             6                   2                   7   

   days_since_prior_order  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0

#### Task #7. Run a check for duplicate values in your df_ords data.

In [618]:
#ID duplicate rows by creating df_dups (which contains only rows that are duplicated)

In [619]:
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [620]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, hour_time_of_order, days_since_prior_order]
Index: []

### #7 Answer:  There are NO DUPLICATE ROWS in df_ords

#### Task #8 Address duplicates using appropriate method. In markdown cell, explain why you used your method of choice.

### #8 Answer: N/A - there are no duplicate rows in df_ords to address

#### Task #9. Export your final, cleaned df_prods & df_ords data as ".csv" files in your "Prepared Data" folder & give them appropriate, succinct names.

### #9 Answer: Exporting final, cleaned dataframes

In [626]:
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared', 'orders_consistency.csv'))

In [627]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared', 'products_consistency.csv'))